File: backend/temp/bus.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import axios from 'axios';
import * as xml2js from 'xml2js';
//주어진 정류장ID와 버스 번호를 사용하여 API에 요청하고, 응답받은 XML 데이터를 처리하여 필요한 정보 반환
//버스 번호에 따른 도착 정보 필터링과 포맷팅을 수행하며, 필요한 경우 오류 처리도 포함
@Injectable()
export class BusService {
  private readonly busRouteMap: { [key: string]: string };

  constructor(private readonly configService: ConfigService) {
    // 환경 변수에서 busRouteMap을 불러와 객체로 변환
    const busRouteMapString = this.configService.get<string>('BUS_ROUTE_MAP');
    this.busRouteMap = JSON.parse(busRouteMapString);
  }

  //주어진 정류장ID와 버스 번호 배열을 사용하여 해당 정류장에서 도착할 버스 정보를 가져오는 메서드
  async getBusArrivalInfo(stationId: string, busNumbers: string[]) {
    const apiKey = this.configService.get<string>('API_KEY'); //환경변수에서 API_KEY 불러오기
    const url = `http://apis.data.go.kr/6410000/busarrivalservice/getBusArrivalList?serviceKey=${apiKey}&stationId=${stationId}`;
    const parser = new xml2js.Parser();

    try {
      const response = await axios.get(url);
      const xmlData = response.data;
      const jsonData = await parser.parseStringPromise(xmlData);

      if (jsonData.response && jsonData.response.msgBody[0].busArrivalList) {
        const items = jsonData.response.msgBody[0].busArrivalList;
        const filteredBusInfo = items
          .filter((item) =>
            Object.values(this.busRouteMap).includes(item.routeId[0]),
          )
          .map((item) => ({
            버스번호: Object.keys(this.busRouteMap).find(
              (key) => this.busRouteMap[key] === item.routeId[0],
            ),
            도착시간: `${item.predictTime1[0]}`,
            남은좌석수:
              item.remainSeatCnt1[0] === '-1'
                ? '정보 없음'
                : `${item.remainSeatCnt1[0]}석 남음`,
          }));
        return filteredBusInfo;
      } else {
        return []; // 데이터가 없을 경우 빈 배열 반환
      }
    } catch (error) {
      return { message: 'Error retrieving bus information', error };
    }
  }
}

File: backend/src/main/java/homerun2/backend/service/BusService.java
package homerun2.backend.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.StringReader;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class BusService {
    private static final Logger logger = LoggerFactory.getLogger(BusService.class);
    private final Map<String, String> busRouteMap;
    private final String apiKey;
    private final RestTemplate restTemplate;
    private final ObjectMapper objectMapper;

    public BusService(
            @Value("${bus.routeMap}") String busRouteMapString,
            @Value("${bus.api-key}") String apiKey) {
        this.apiKey = apiKey;
        this.restTemplate = new RestTemplate();
        this.objectMapper = new ObjectMapper();
        this.busRouteMap = parseBusRouteMap(busRouteMapString);
        logger.info("Initialized BusService with routeMap: {}", busRouteMap);
    }

    private Map<String, String> parseBusRouteMap(String busRouteMapString) {
        try {
            return objectMapper.readValue(busRouteMapString, new TypeReference<Map<String, String>>() {
            });
        } catch (Exception e) {
            logger.error("Failed to parse bus route map: {}", busRouteMapString, e);
            throw new RuntimeException("Failed to parse bus route map", e);
        }
    }

    public List<Map<String, String>> getBusArrivalInfo(String stationId, List<String> busNumbers) {
        try {
            // URL을 직접 구성
            String baseUrl = "http://apis.data.go.kr/6410000/busarrivalservice/getBusArrivalList";
            String url = String.format("%s?serviceKey=%s&stationId=%s",
                    baseUrl,
                    URLEncoder.encode(apiKey, StandardCharsets.UTF_8),
                    stationId);

            logger.info("Requesting bus arrival info from URL: {}", url);
            logger.info("Station ID: {}, Bus Numbers: {}", stationId, busNumbers);

            String xmlResponse = restTemplate.getForObject(new URI(url), String.class);
            logger.info("Received XML response: {}", xmlResponse);

            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            Document doc = builder.parse(new InputSource(new StringReader(xmlResponse)));

            // Check for error response
            NodeList errorHeader = doc.getElementsByTagName("cmmMsgHeader");
            if (errorHeader.getLength() > 0) {
                Element header = (Element) errorHeader.item(0);
                String errorMsg = getTextContent(header, "errMsg");
                String authMsg = getTextContent(header, "returnAuthMsg");
                String reasonCode = getTextContent(header, "returnReasonCode");

                logger.error("API Error - Message: {}, Auth: {}, Code: {}", errorMsg, authMsg, reasonCode);
                Map<String, String> errorMap = new HashMap<>();
                errorMap.put("message", "API Error: " + errorMsg);
                errorMap.put("details", authMsg);
                return List.of(errorMap);
            }

            // Check regular response status
            NodeList headerList = doc.getElementsByTagName("msgHeader");
            if (headerList.getLength() > 0) {
                Element header = (Element) headerList.item(0);
                String resultCode = getTextContent(header, "resultCode");
                String resultMessage = getTextContent(header, "resultMsg");
                logger.info("API Response - Code: {}, Message: {}", resultCode, resultMessage);

                if (!"0".equals(resultCode)) {
                    throw new RuntimeException("API Error: " + resultMessage);
                }
            }

            NodeList busArrivalList = doc.getElementsByTagName("busArrivalList");
            List<Map<String, String>> filteredBusInfo = new ArrayList<>();

            logger.info("Found {} bus arrival entries", busArrivalList.getLength());

            for (int i = 0; i < busArrivalList.getLength(); i++) {
                Element item = (Element) busArrivalList.item(i);
                String routeId = getTextContent(item, "routeId");
                logger.debug("Processing routeId: {}", routeId);

                if (busRouteMap.containsValue(routeId)) {
                    Map<String, String> busInfo = new HashMap<>();
                    String busNumber = getBusNumberByRouteId(routeId);
                    String predictTime = getTextContent(item, "predictTime1");
                    String remainSeatCnt = getTextContent(item, "remainSeatCnt1");

                    busInfo.put("버스번호", busNumber);
                    busInfo.put("도착시간", predictTime);
                    busInfo.put("남은좌석수", remainSeatCnt.equals("-1") ? "정보 없음" : remainSeatCnt + "석 남음");

                    filteredBusInfo.add(busInfo);
                    logger.debug("Added bus info: {}", busInfo);
                }
            }
            return filteredBusInfo;
        } catch (Exception e) {
            logger.error("Error retrieving bus information", e);
            Map<String, String> errorMap = new HashMap<>();
            errorMap.put("message", "Error retrieving bus information");
            errorMap.put("error", e.getMessage());
            return List.of(errorMap);
        }
    }

    private String getTextContent(Element element, String tagName) {
        NodeList nodeList = element.getElementsByTagName(tagName);
        if (nodeList.getLength() > 0) {
            return nodeList.item(0).getTextContent();
        }
        return "";
    }

    private String getBusNumberByRouteId(String routeId) {
        return busRouteMap.entrySet()
                .stream()
                .filter(entry -> entry.getValue().equals(routeId))
                .map(Map.Entry::getKey)
                .findFirst()
                .orElse("");
    }
}
